### PR TITLE
Chore: Makefile / Magefile task to build plugins

### PR DESCRIPTION
**Pull request by andresmgot**
**What is this feature?**

Needs https://github.com/grafana/grafana/pull/76816

Allow to build binaries for all decoupled plugins. Decoupled plugins are identified for the ones having a `main.go` (if no specific plugin is manually set). To build plugins (currently only `grafana-testdata-datasource`) execute:

```bash
# build all plugins
$ make build-plugins-go

# build a single plugin
$ make build-plugin-go PLUGIN_ID=grafana-testdata-datasource
```

**Which issue(s) does this PR fix?**:

Needed for #74004

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/tsdb/Magefile.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"path/filepath"
+	"strings"
+
+	build "github.com/grafana/grafana-plugin-sdk-go/build"
+)
+
+func find(dir string, name string) ([]string, error) {
+	files := []string{}
+	err := filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
+		if filepath.Base(path) == name {
```
Comment (xnyo) [inline pkg/tsdb/Magefile.go:17]: Nit: You have to handle `err` inside the `WalkFunc` here:

```suggestion
		if err != nil {
			return err
		}
		if filepath.Base(path) == name {
```

File: pkg/tsdb/Magefile.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	if err != nil {
+		return "", err
+	}
+	if len(pluginJSONMatches) == 0 {
+		return "", fmt.Errorf("Could not find plugin.json")
+	}
+	pluginJSONPath := ""
+	for _, pluginJSONMatch := range pluginJSONMatches {
+		// Ignore dist folder
+		if !strings.Contains(pluginJSONMatch, "dist") {
```
Comment (xnyo) [inline pkg/tsdb/Magefile.go:37]: Nit: Maybe it's more robust to check the folder that contains `plugin.json`, and checking that that part is `dist` instead?

```suggestion
		if filepath.Dir(pluginJSONMatch) != "dist" {
```

File: pkg/tsdb/Magefile.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	buildPlugin(rootDir, pluginJSONDir)
+	return nil
+}
+
+func BuildAllPlugins() error {
+	// Plugins need to have a main.go file
+	matches, err := find(".", "main.go")
+	if err != nil {
+		panic(err)
```
Comment (xnyo) [inline pkg/tsdb/Magefile.go:93]: Nit:
```suggestion
		return err
```

File: pkg/tsdb/Magefile.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	for _, match := range matches {
+		// Get the directory name of the plugin
+		pluginDir := strings.Split(match, fmt.Sprintf("%c", os.PathSeparator))[0]
```
Comment (xnyo) [inline pkg/tsdb/Magefile.go:97]: Nit:
```suggestion
		parts := filepath.SplitList(match)
		if len(parts) == 0 {
			continue
		}
		pluginDir := parts[0]
```

**Review by xnyo — COMMENTED** [review]
Awesome work :clap::clap:

Leaving a couple of comments but looks good to me so far!

File: pkg/tsdb/Magefile.go (hunk 4 above)
Comment (andresmgot) [reply pkg/tsdb/Magefile.go:97]: SplitList does a much less useful thing: https://pkg.go.dev/path/filepath#SplitList :smile:

**Review by marefr — APPROVED** [review]
LGTM

**Review by xnyo — APPROVED** [review]
LGTM!

**Pull request merged**